### PR TITLE
Further apply read-only txn hints

### PR DESCRIPTION
**Pull request by achimnol**

### CODE REVIEW

File: src/ai/backend/manager/api/auth.py [hunk 1]
```diff
@@ -437,12 +438,12 @@ async def _query_cred():
...
                 raise AuthorizationFailed('Access key not found')
 
-            async def _pipe_builder(r: Redis):
+            async def _pipe_builder(r: Redis) -> RedisPipeline:
                 pipe = r.pipeline()
                 num_queries_key = f'kp:{access_key}:num_queries'
                 pipe.incr(num_queries_key)
                 pipe.expire(num_queries_key, 86400 * 30)  # retention: 1 month
-                await pipe.execute()
+                return pipe
```
Comment (fregataa) [inline src/ai/backend/manager/api/auth.py:446]: I am just curious.
is this a big change or difference?

File: src/ai/backend/manager/api/session.py [hunk 2]
```diff
@@ -879,117 +884,109 @@ async def create_cluster(request: web.Request, params: Any) -> web.Response:
...
-                    target = git['repository'].split('/')[-1]
+        if template['spec']['sess_type'] == 'interactive':
+            kernel_config['sess_type'] = SessionTypes.INTERACTIVE
+        elif template['spec']['sess_type'] == 'batch':
+            kernel_config['sess_type'] = SessionTypes.BATCH
+
+        # TODO: Remove `type: ignore` when mypy supports type inference for walrus operator
+        # Check https://github.com/python/mypy/issues/7316
+        # TODO: remove `NOQA` when flake8 supports Python 3.8 and walrus operator
+        # Check https://gitlab.com/pycqa/flake8/issues/599
```
Comment (kyujin-cho) [inline src/ai/backend/manager/api/session.py:928]: Though this is not related to, since this PR is changing this code block: both python/mypy#7316 and PyCQA/flake8#323 (migrated from Gitlab issue) is both resolved and released, I think it'll be good to remove all `#noqa` annotations.

File: src/ai/backend/manager/api/session.py [hunk 3]
```diff
@@ -879,117 +884,109 @@ async def create_cluster(request: web.Request, params: Any) -> web.Response:
...
+            if bootstrap := runtime_opt['bootstrap']:
+                kernel_config['bootstrap_script'] = bootstrap
+            if startup := runtime_opt['startup_command']:
+                kernel_config['startup_command'] = startup
+
+        if resources := template['spec'].get('resources'):
+            kernel_config['creation_config']['resources'] = resources
+
+        if git := template['spec']['kernel']['git']:  # noqa
+            if _dest := git.get('dest_dir'):  # noqa
```
Comment (kyujin-cho) [inline src/ai/backend/manager/api/session.py:941]:
```suggestion
        if git := template['spec']['kernel']['git']:
            if _dest := git.get('dest_dir'):
```

File: src/ai/backend/manager/api/session.py [hunk 4]
```diff
@@ -879,117 +884,109 @@ async def create_cluster(request: web.Request, params: Any) -> web.Response:
...
+                cmd_builder += git['repository']
+            if branch := git.get('branch'):  # noqa
+                cmd_builder += f' -b {branch}'
+            cmd_builder += f' {target}\n'
+
+            if commit := git.get('commit'):  # noqa
+                cmd_builder = 'CWD=$(pwd)\n' + cmd_builder
+                cmd_builder += f'cd {target}\n'
+                cmd_builder += f'git checkout {commit}\n'
+                cmd_builder += 'cd $CWD\n'
```
Comment (kyujin-cho) [inline src/ai/backend/manager/api/session.py:960]:
```suggestion
            if credential := git.get('credential'):
                proto, url = git['repository'].split('://')
                cmd_builder += f'{proto}://{credential["username"]}:{credential["password"]}@{url}'
            else:
                cmd_builder += git['repository']
            if branch := git.get('branch'):
                cmd_builder += f' -b {branch}'
            cmd_builder += f' {target}\n'

            if commit := git.get('commit'):
                cmd_builder = 'CWD=$(pwd)\n' + cmd_builder
                cmd_builder += f'cd {target}\n'
                cmd_builder += f'git checkout {commit}\n'
                cmd_builder += 'cd $CWD\n'
```

**Review by kyujin-cho — CHANGES_REQUESTED** [review]

File: src/ai/backend/manager/api/session.py (hunk 2 above)
Comment (achimnol) [reply src/ai/backend/manager/api/session.py:928]: Good point! 👍🏼

**Review by kyujin-cho — APPROVED** [review]
LGTM!
